Add InitialData base class, namespace, and library

yoonso0-0 opened the pull request:
## Proposed changes

A fraction (first 3 commits) of the draft PR #3448.

* Add Initial Data group to Groupdefs
* Define an `InitialData` abstract base class in `PointwiseFunctions/InitialDataUtilities/InitialData.hpp` which will have analytic solutions and analytic data classes as its derived classes. Another derived class named `NumericInitialDataFromFile`, for importing numerc initial data, needs to be added in the upcoming PRs.
* All added structs and classes are in a new namespace `InitialDataUtilities` 
* Added a new library `InitialDataUtilities`.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

yoonso0-0 committed:
Add forward declarations to Evolution/ComputeTags.hpp

yoonso0-0 committed:
Rename initialization_option_tags in SetVariables.hpp

nilsvu commented on src/PointwiseFunctions/InitialDataUtilities/InitialData.hpp:
```diff
@@ -0,0 +1,28 @@
+ * \ingroup InitialDataGroup
+ * \brief The abstract base class for initial data of evolution systems.
+ */
+class InitialData : public PUP::able {
```
I think let's put this in the `evolution` namespace. It's evolution-specific and it's a lot more intuitive to write `evolution::InitialData` and `evolution::Tags::InitialData`.

yoonso0-0 replied:
And remove `InitialDataUtilities` namespace, do you mean?

nilsvu replied:
We may need it later, so it's up to you if you want to leave it in or add when it's needed. I'd remove it for now.

yoonso0-0 replied:
It seems that throughout the code in general the namespace follows the name of the library, so I think it'd be better keep `InitialDataUtilites` since we add a new library here -- if we need to mark something as evolution-specific, how about putting these into `InitialDataUtilities::evolution`?

nilsdeppe replied:
I agree with Yoonsoo. Placing something in the `evolution` namespace here is rather unusual and not what I had in my notes from the conversation the three of us had.

nilsvu replied:
ok. Don't you think `InitialDataUtilities::InitialData` is quite clunky, and it would be nicer to make this `evolution::InitialData`? I suggested just keeping files/libs where they are and adjust the namespace, but maybe there's a better way.

nilsdeppe replied:
That sort of adds a cyclic dependency. I personally like having each library have its own namespace, with "sub libraries" being nested. Moving `InitialData` to `src/Evolution` would create cyclic dependencies. Leaving it here would avoid the technical dependency, but still have a conception cyclic dependency. That's why I'd rather have the "clunky" name

nilsvu commented on src/PointwiseFunctions/InitialDataUtilities/Tags/InitialData.hpp:
```diff
@@ -0,0 +1,46 @@
+namespace Tags {
+// Base tag for the InitialData struct. Can be used to retrieve the InitialData
+// type object from the cache without having to know its template parameter.
+struct InitialDataBase : db::BaseTag {};
```
I believe we don't need the base tag anymore based on the plan we made. You can just use the `InitialData` base class as the type of the simple tag and remove its template parameter.

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -485,6 +485,11 @@
  */
 
+/*!
+ * \defgroup InitialDataGroup Initial Data
```
Do we really need or want another group for this? What's going to go in this group? If the initial data stuff is mostly geared towards evolution, I'd say put the `InitialDataUtilities` namespace in the `EvolutionSystemsGroup` :shrug: It's not clear to me what the right place is, but having a separate thing for just one class seems to me to be the wrong thing to do.

nilsdeppe commented on src/PointwiseFunctions/InitialDataUtilities/InitialData.hpp:
```diff
@@ -0,0 +1,28 @@
+
+#include "Parallel/CharmPupable.hpp"
+
+namespace InitialDataUtilities {
```
Please document the namespace

nilsdeppe commented on src/PointwiseFunctions/InitialDataUtilities/Tags/InitialData.hpp:
```diff
@@ -0,0 +1,46 @@
+#include "Utilities/TMPL.hpp"
+
+namespace InitialDataUtilities {
+namespace OptionTags {
```
Please document the namespace

nilsdeppe commented on src/PointwiseFunctions/InitialDataUtilities/Tags/InitialData.hpp:
```diff
@@ -0,0 +1,46 @@
+namespace InitialDataUtilities {
+namespace OptionTags {
+/*!
+ * \ingroup OptionGroupsGroup
```
This is the wrong group. `InitialData` is a tag, not a group

nilsdeppe commented on src/PointwiseFunctions/InitialDataUtilities/Tags/InitialData.hpp:
```diff
@@ -0,0 +1,46 @@
+};
+}  // namespace OptionTags
+
+namespace Tags {
```
Please document the namespace

nilsdeppe commented on src/Evolution/Initialization/SetVariables.hpp:
```diff
@@ -52,8 +52,7 @@ namespace Actions {
 ///   * System::primitive_variables_tag (if system has primitive variables)
```
Can you make these types of commit messages more specific? Eg `Rename initialization_option_tags in SetVariables.hpp` so that it's easier to figure out if the commit in the history could be responsible for a bug

nilsdeppe commented on src/Evolution/ComputeTags.hpp:
```diff
@@ -13,6 +13,13 @@
 #include "Time/Tags.hpp"
```
Commit message, maybe `Add forward decls to Evolution/ComputeTags`?

nilsdeppe commented:
My biggest worry is how this renders and shows up in the documentation. My guess is that currently it doesn't really. E.g. the namespaces aren't documented, and it's not clear what the best way to sort these structures into doxygen groups.

yoonso0-0 commented:
@nilsdeppe @nilsleiffischer  My suggestion is to move existing `AnalyticDataGroup` and `AnalyticSolutionGroup` into the new `InitialData` group (which will require additional work than the current PR). I think it'd be better keep `EvolutionSystemsGroup` for listing each systems, and place related utilities at somewhere else.

nilsdeppe commented:
I don't like moving the analytic data and solutions group into the `InitialData` group. I guess `AnalyticData` is really only ever initial data, but analytic solutions are not just initial data.

Things I find a bit weird right now (but I haven't made up my mind on yet):
1. having `InitialData` base class in the `evolution` namespace because then the `evolution` namespace is used outside of `src/Evolution`
2. the existence of the `AnalyticDataGroup` and `AnalyticSolutionGroup`. The solutions are in `System::solutions` and `System::AnalyticData`, which I think is where they should be in the documentation.
3. maybe we should have `evolution::initial_data` as a namespace and put the base class in there, as well as future utilities?
4. if we remove the `AnalyticSolution/DataGroup` then where do we document `is_analytic_data_v`? Well, those metafunctions are already in the `evolution` namespace so we can just leave them there, I guess. These metafunctions are in the `src/Evolution` directory
 The`MarkAsBlah` aren't documented, actually. 

I'll need to think about it a bit and see what @kidder and @wthrowe have to say.

nilsvu commented:
@nilsdeppe I like your suggestion for an `evolution::initial_data` namespace. The utilities you mention in `4.` can also go in there.

nilsdeppe commented:
Yea, that's the best solution. So I think the specific changes to this PR then are (@nilsleiffischer please confirm so @yoonso0-0 doesn't do work and then undo it :) ):
- remove `InitialDataGroup`. We'll need some general documentation for the `evolution` namespace/code anyway. Probably some sort of a markdown file, but that's beyond the scope of this PR. It'll depend on how we structure tutorials, etc.
- switch `InitialDataUtilities` to `evolution::initial_data` namespace
- document `evolution::initial_data` namespace

We can move the utilities in the future once we have a better feeling for this structure.

nilsvu commented:
Sounds good 👍

yoonso0-0 commented:
Alright, I'll put these into `evolution::initial_data` namespace and push. Should we keep all source files still in `PointwiseFunctions/InitialDataUtilities` with the new library `InitialDataUtilities`?

nilsdeppe commented:
I'm fine with leaving the directory and library name as-is. I can't think of anything better. @nilsleiffischer ?

nilsdeppe commented on src/PointwiseFunctions/InitialDataUtilities/InitialData.hpp:
```diff
@@ -0,0 +1,33 @@
+ */
+namespace initial_data {
+/*!
+ * \brief The abstract base class for initial data of evolution systems.
```
Maybe add a comment like `All analytic solutions and analytic data must virtually inherit from this class`? I think we discussed that they probably all need virtual inheritance. I expect this will be non-intuitive and so should be documented.

nilsvu replied:
Are you sure this is needed? Should only be needed if we have solution/data classes that inherit from multiple other solution/data classes, I can't think of one.

nilsdeppe replied:
Well the recommendation is that all analytic data and solutions inherit from this. There are analytic data that inherit from analytic solutions, which should also explicitly inherit from this base class since the inheritance of the analytic solution is typically private. I believe that leaves the base classes of the analytic solution also private. Public inheritance of the solution is definitely incorrect and will cause problems.

nilsdeppe reviewed (commented):
Overall LGTM, just one request on clarifying the documentation a bit more. Please squash anything immediately :)

yoonso0-0 commented:
@nilsdeppe @nilsleiffischer Namespace `evolution::initial_data` makes conflict with the existing `evolution::initial_data()` function in `Evolution/Initialization/InitialData.hpp` which makes gcc fail (see the CI failure at #3600). May I change the namespace to `evolution::InitialData` and squash in this PR?

nilsdeppe commented:
I think it makes more sense to more the `initial_data` function into `evolution::Initialization` :shrug:

yoonso0-0 commented:
Rebased and squashed. Thanks very much for the review :) This was a hard job.

nilsvu commented on src/Evolution/Initialization/InitialData.hpp:
```diff
@@ -9,7 +9,7 @@
 #include "Utilities/TMPL.hpp"
 
-namespace evolution {
+namespace evolution ::initial_data {
```
I believe @nilsdeppe meant to make this `evolution::Initialization`. There are a few places in the code that use this function, that you probably have to change to use the new namespace. (squash directly)

nilsdeppe replied:
yes, that's what I meant. Also, please clang-format (no space before `::`)

yoonso0-0 replied:
Ooops. yep, that's right. My bad :joy:

nilsdeppe commented on src/Evolution/Initialization/InitialData.hpp:
```diff
@@ -9,7 +9,7 @@
 #include "Utilities/TMPL.hpp"
 
-namespace evolution {
+namespace evolution::Initialization {
```
this should be  a separate commit before adding the base class

yoonso0-0 committed:
Add a new initial data base class

- Define InitialData base class in evolution::initial_data namespace
- Add InitialDataUtilities library

yoonso0-0 committed:
Move evolution::initial_data() to sub-namespace
